sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/NetFrameworkVersionProvider.cs

```diff
@@ -57,20 +57,50 @@ public NetFrameworkVersion GetDotNetFrameworkVersion(Compilation compilation)
                 return NetFrameworkVersion.Unknown;
             }
 
-            if (!debuggerConstructorSymbol.GetAttributes().Any(attribute => attribute.AttributeClass.Name.Equals("ObsoleteAttribute")))
+            if (Probably35(debuggerConstructorSymbol))
             {
-                // the constructor was still not deprecated in .NET Framework 3.5
                 return NetFrameworkVersion.Probably35;
             }
-
-            var typeSymbol = mscorlibAssembly.GetTypeByMetadataName("System.IO.UnmanagedMemoryStream");
-            // this was not present in .NET Framework 4.5.1 and became present in 4.5.2
-            if (typeSymbol != null && !typeSymbol.GetMembers("FlushAsync").IsEmpty)
+            else if (AtLeast46(mscorlibAssembly))
+            {
+                return NetFrameworkVersion.After46;
+            }
+            else if (AtLeast452(mscorlibAssembly))
             {
                 return NetFrameworkVersion.After452;
             }
-
-            return NetFrameworkVersion.Between4And451;
+            else
+            {
+                return NetFrameworkVersion.Between4And451;
+            }
         }
+
+        /// <remarks>the constructor was still not deprecated in .NET Framework 3.5</remarks>
+        private static bool Probably35(ISymbol debuggerConstructorSymbol)
+            => !debuggerConstructorSymbol
+            .GetAttributes()
+            .Any(attribute => attribute.AttributeClass.Name.Equals("ObsoleteAttribute"));
+
+        /// <summary>Returns true if the .NET version was at least 4.5.2.</summary>
+        /// <remarks>
+        /// Detected by checking the existence of System.IO.UnmanagedMemoryStream.FlushAsync
+        /// introduced in .NET 4.5.2
+        /// </remarks>
+        private static bool AtLeast452(IAssemblySymbol mscorlibAssembly)
+            => mscorlibAssembly
+            .GetTypeByMetadataName("System.IO.UnmanagedMemoryStream")?
+            .GetMembers("FlushAsync")
+            .Any() == true;
+
+        /// <summary>Returns true if the .NET version was at least 4.6.</summary>
+        /// <remarks>
+        /// Detected by checking the existence of System.Array.Empty&lt;T&gt;,
+        /// introduced in .NET 4.6
+        /// </remarks>
+        private static bool AtLeast46(IAssemblySymbol mscorlibAssembly)
+            => mscorlibAssembly
+            .GetTypeByMetadataName("System.Array")?
+            .GetMembers("Empty")
+            .Any() == true;
```

Comment:
Same from the copy/pasted original above
```suggestion
            .Any();
```

Reply:
Nope, `GetTypeByMetadataName()?` will return bool? for Any()

Reply:
Ah, ok. I missed that. Can you revert it then? As it doesn't compile now.